mreg_cli/utilities/api.py

```diff
@@ -0,0 +1,270 @@
+        _update_token(user, password)
+    except CliError as e:
+        e.print_self()
+
+
+def logout() -> None:
+    """Logout from MREG."""
+    path = requests.compat.urljoin(mregurl, "/api/token-logout/")
+    # Try to logout, and ignore errors
+    try:
+        session.post(path)
+    except requests.exceptions.ConnectionError:
+        pass
+
+
+def update_token() -> None:
+    """Update the token."""
+    password = prompt("You need to re-autenticate\nEnter password: ", is_password=True)
+    try:
+        _update_token(username, password)
+    except CliError as e:
+        e.print_self()
+
+
+def _update_token(username: str, password: str) -> None:
+    """Perform the actual token update."""
+    tokenurl = requests.compat.urljoin(mregurl, "/api/token-auth/")
+    try:
+        result = requests.post(tokenurl, {"username": username, "password": password})
+    except requests.exceptions.ConnectionError as err:
+        error(err)
+    except requests.exceptions.SSLError as e:
+        error(e)
+    if not result.ok:
+        try:
+            res = result.json()
+        except json.JSONDecodeError:
+            res = result.text
+        if result.status_code == 400:
+            if "non_field_errors" in res:
+                cli_error("Invalid username/password")
+        else:
+            cli_error(res)
+    token = result.json()["token"]
+    session.headers.update({"Authorization": f"Token {token}"})
+    try:
+        with open(mreg_auth_token_file, "w", encoding="utf-8") as tokenfile:
+            tokenfile.write(f"{username}¤{token}")
+    except FileNotFoundError:
+        pass
+    except PermissionError:
+        pass
+    set_file_permissions(mreg_auth_token_file, 0o600)
+
+
+def result_check(result: "ResponseLike", operation_type: str, url: str) -> None:
+    """Check the result of a request."""
+    if not result.ok:
+        message = f'{operation_type} "{url}": {result.status_code}: {result.reason}'
+        try:
+            body = result.json()
+        except ValueError:
+            pass
+        else:
+            message += "\n{}".format(json.dumps(body, indent=2))
+        cli_warning(message)
+
+
+def _request_wrapper(
+    operation_type: str,
+    path: str,
+    params: Dict[str, str] = None,
+    ok404: bool = False,
+    first: bool = True,
+    use_json: bool = False,
+    **data: Any,
+) -> Optional["ResponseLike"]:
+    """Wrap request calls to MREG for logging and token management."""
+    if params is None:
+        params = {}
+    url = requests.compat.urljoin(mregurl, path)
+
+    if use_json:
+        result = getattr(session, operation_type)(url, json=params, timeout=HTTP_TIMEOUT)
+    else:
+        result = getattr(session, operation_type)(
+            url, params=params, data=data, timeout=HTTP_TIMEOUT
+        )
+    result = cast(requests.Response, result)  # convince mypy that result is a Response
+
+    OutputManager().recording_request(operation_type, url, params, data, result)
```

Comment:
The `params` parameter of `OutputManager.recording_request` takes a `str` arg, yet we pass it a `dict` here?

Reply:
This seems to have been resolved at some point?

Reply:
Must have been a result of the type annotation fixes in my PR.